tensorflow/lite/kernels/internal/common.h

```diff
@@ -156,6 +156,27 @@ inline int32 MultiplyByQuantizedMultiplier(int32 x, int32 quantized_multiplier,
                              right_shift);
 }
 
+inline int32 MultiplyByQuantizedMultiplier(std::int64_t x,
+                                           int32 quantized_multiplier,
+                                           int shift) {
+  // Inputs:
+  // - quantized_multiplier has fixed point at bit 31
+  // - shift is -31 to +7 (negative for right shift)
+  //
+  // Assumptions: The following input ranges are assumed
+  // - quantize_scale>=0  (the usual range is (1<<30) to (1>>31)-1)
+  // - scaling is chosen so final scaled result fits in int32
+  // - input x is in the range -(1<<47) <= x < (1<<47)
```

Comment:
Just for my understanding, what's the interaction between this range and the shift?

y = x * multiplier * 2 ^ (-31 + shift) and multiplier is [2^30, 2^31). If shift is as high as 7, multiplier * 2 ^ (-31 + shift) is within [2^6, 2^7). Multiplied to x, where -(1<<47) <= x < (1<<47), it seems possible to overflow the int32 output.

Did I miss anything? Very likely it works for real use cases (as shown in the test cases) but just want to double check. Thanks.

Reply:
Hi. Yes, you are quite correct that if a high shift value is chosen then this could overflow the 32-bit result. There is an assumption that quantize_multiplier and shift values are chosen by the caller so that the result cannot overflow (listed in the assumptions comment).

In practical use, the quantize_multipler and shift will be chosen to rescale to the scaling value of the output tensor. Therefore the expected range of values after the scaling is 16-bits. To overflow the 32-bit result, I think the model quantization would have had to produce a scale for the output tensor which is too small for the tensor contents by a factor of (1<<16). Such a scale would seem to cause the model to be unusable in any event. So, we don't think this is an issue in practice.

I think there is a similar assumption in the version of MultiplyByQuantizedMultiplier() with 32-bit input 'x' on line 154 above where the intermediate value (x * (1 << left_shift)) could overflow if a large left_shift is used.

Reply:
@jianlijianli, can you please check Dominic's explanation and let us know how we can move this PR forward. Thanks!

Reply:
Thanks for the reply. Yes, that is my understanding as well. Just wanted to make sure.